frame/executive/src/lib.rs

```diff
@@ -366,6 +380,19 @@ where
 		sp_tracing::enter_span!( sp_tracing::Level::TRACE, "finalize_block" );
 		<frame_system::Module<System>>::note_finished_extrinsics();
 		let block_number = <frame_system::Module<System>>::block_number();
+
+
+		let weight =  <frame_system::Module<System>>::block_weight();
+		let max_weight =  <System::BlockWeights as frame_support::traits::Get<_>>::get().max_block;
+		let mut remaining_weight = max_weight.saturating_sub(weight.total());
+
+		if remaining_weight > 0 {
+			let mut reduced_weight = <frame_system::Module<System> as OnIdle<System::BlockNumber>>::on_idle(block_number, remaining_weight);
+			remaining_weight = remaining_weight.saturating_sub(reduced_weight);
+			reduced_weight = <AllModules as OnIdle<System::BlockNumber>>::on_idle(block_number, remaining_weight).saturating_add(reduced_weight);
+			<frame_system::Module::<System>>::register_extra_weight_unchecked(reduced_weight, DispatchClass::Mandatory);
+		}
+
```

Comment:
seems like enough duplication to extract out into a function? (probably including `on_finalize` calls below)